internal/safebuffer/safebuffer_test.go

```diff
@@ -0,0 +1,203 @@
+package safebuffer
+
+import (
+	"context"
+	"io"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func bufWrite(buf *Buffer, v string) <-chan error {
+	ch := make(chan error)
+	go func() {
+		_, err := buf.Write([]byte(v))
+		ch <- err
+	}()
+	return ch
+}
+
+func TestSafeBuffer(t *testing.T) {
+	t.Parallel()
+
+	t.Run("sync", func(t *testing.T) {
+		t.Parallel()
+		assert := assert.New(t)
+		require := require.New(t)
+		ctx := context.Background()
+
+		jobDone := make(chan struct{})
+		buf := New(ctx, jobDone)
+
+		var err error
+		for range 3 {
+			if e := <-bufWrite(buf, "foo"); e != nil {
+				err = e
+				break
+			}
+		}
+
+		require.NoError(err)
+		assert.Equal("foofoofoo", buf.buf.String())
+
+		r0 := buf.NewReader()
+		b := make([]byte, 3)
+		var n int
+		for range 3 {
+			n, err = r0.Read(b)
+			require.NoError(err)
+			assert.Equal(3, n)
+		}
+
+		done := make(chan struct{})
+		go func() {
+			defer close(done)
+			for err != io.EOF {
+				n, err = r0.Read(b)
+			}
+		}()
+
+		timer := time.NewTimer(10 * time.Millisecond)
+		select {
+		case <-timer.C:
+		case <-done:
+			t.Fatal("expected Read not to complete because jobDone wasn't closed")
+		}
+
+		close(jobDone)
+
+		timer.Reset(10 * time.Millisecond)
+		select {
+		case <-timer.C:
+			t.Fatal("expected Read to complete because jobDone was closed")
+		case <-done:
+		}
+		assert.Equal(io.EOF, err)
+		assert.Equal(0, n)
```

Comment:
What is this block testing? `done` is already closed and `n` and `err` are leftovers, or am I missing something?

Reply:
It's checking to make sure that the reader blocks until `io.EOF` is received. Here's the sequence:
1. Write to the buffer
2. Synchronous read everything from the buffer
3. Try reading more from the buffer in the goroutine
4. This times out because nothing more is written
5. Close the `jobDone` channel to simulate the job completing
6. Now the reader should should return `io.EOF` and the `done` channel will close

I'll clarify in the code